### PR TITLE
add Half support for sinh, cosh, ploygamma, entr and i0e on CPU

### PR DESCRIPTION
**Pull request by jiayisunx**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #99002

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

**Review by mingfeima — APPROVED** [review]

File: aten/src/ATen/native/cpu/UnaryOpsKernel.cpp [hunk 1]
```diff
@@ -460,12 +460,23 @@ static void nan_to_num_kernel(
...
-    cpu_kernel(iter, [=](scalar_t a){
-        return calc_i0(static_cast<scalar_t>(beta) * std::sqrt(1 - std::pow((a - alpha) / alpha, static_cast<scalar_t>(2.0)))) / calc_i0(static_cast<scalar_t>(beta));
+  if (at::isReducedFloatingType(iter.dtype())) {
+    AT_DISPATCH_REDUCED_FLOATING_TYPES(iter.dtype(), "kaiser_window_cpu", [&]() {
+      const float alpha = static_cast<float>((window_length - 1) / 2.0);
+      const float beta_ = static_cast<float>(beta);
+      cpu_kernel(iter, [=](scalar_t a){
+          return calc_i0(beta_ * std::sqrt(1 - std::pow((static_cast<float>(a) - alpha) / alpha, static_cast<float>(2.0)))) / calc_i0(beta_);
+      });
     });
```
Comment (jgong5) [inline aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:470]: Can we combine the implementation with float and use the `opmath_t<scalar_t>` for intermediate dtype?

**Review by jgong5 — APPROVED** [review]

**Comment by jiayisunx** [comment]
@ngimel, can you please review this PR? thanks!

**Comment by jiayisunx** [comment]
@ngimel

**Comment by jiayisunx** [comment]
@ngimel, could you please help review this PR? thanks!

**Review by ngimel — APPROVED** [review]

**Comment by osalpekar** [comment]
I'll need to revert this in order to cleanly revert https://github.com/pytorch/pytorch/pull/101976, which breaks certain jobs. We can check this one back in once the offending commits have been reverted.

**Comment by osalpekar** [comment]
@jiayisunx I reverted the breaking commits that your change depended on, so this PR no longer cleanly patches on main as-is. Do you mind rebasing this PR and then kicking off the merge again? Apologies for the inconvenience - we had to revert earlier commits in order to keep trunk gree.

**Pull request closed without merging**